File: aip/packages/typescript/src/analysis/agreement.ts
/**
 * Card-Conscience Agreement validation for the Agent Integrity Protocol.
 *
 * Validates that conscience values are compatible with the Alignment Card
 * at initialization time. If conflicts are found, initialization MUST fail.
 */

import type { AlignmentCard } from "../schemas/config.js";
import type { ConscienceValue } from "../schemas/conscience.js";
import type { CardConscienceAgreement } from "../schemas/agreement.js";

/** Escape special regex characters in a string */
function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/** Test if text contains the term as a whole word (not as a substring of another word) */
function containsWholeWord(text: string, term: string): boolean {
  return new RegExp(`\\b${escapeRegex(term)}\\b`).test(text);
}

/**
 * Validate that conscience values are compatible with the alignment card.
 * Called at AIP initialization — throws if conflicts are found.
 *
 * Conflict detection:
 * - A BOUNDARY value that contradicts a bounded_action (e.g., BOUNDARY says
 *   "never write files" but bounded_actions includes "write_files")
 *
 * Augmentation detection:
 * - A BOUNDARY/FEAR value that reinforces an escalation_trigger
 * - A BOUNDARY value that matches a forbidden_action
 */
export function validateAgreement(
  card: AlignmentCard,
  conscienceValues: ConscienceValue[]
): CardConscienceAgreement {
  const conflicts: CardConscienceAgreement["conflicts"] = [];
  const augmentations: CardConscienceAgreement["augmentations"] = [];

  for (const value of conscienceValues) {
    // Only check BOUNDARY and FEAR types (others are not injected)
    if (value.type !== "BOUNDARY" && value.type !== "FEAR") continue;

    const contentLower = value.content.toLowerCase();

    // Check for conflicts: BOUNDARY contradicts bounded_actions
    if (value.type === "BOUNDARY" && card.autonomy_envelope.bounded_actions) {
      for (const action of card.autonomy_envelope.bounded_actions) {
        // Use word-boundary matching to avoid false positives
        // (e.g., "execute" in BOUNDARY text should not match action "exec")
        if (containsWholeWord(contentLower, action.toLowerCase().replace(/_/g, " ")) ||
            containsWholeWord(contentLower, action.toLowerCase())) {
          // Check if the BOUNDARY says "never" or "no" + the action
          // This is a potential conflict — card allows it, conscience forbids it
          if (contentLower.includes("never") || contentLower.includes("no ") || contentLower.includes("don't") || contentLower.includes("do not")) {
            conflicts.push({
              conscience_value: value,
              card_field: "autonomy_envelope.bounded_actions",
              description: `BOUNDARY "${value.content}" conflicts with bounded action "${action}"`,
            });
          }
        }
      }
    }

    // Check for augmentations: value reinforces forbidden_actions
    if (card.autonomy_envelope.forbidden_actions) {
      for (const action of card.autonomy_envelope.forbidden_actions) {
        if (containsWholeWord(contentLower, action.toLowerCase().replace(/_/g, " ")) ||
            containsWholeWord(contentLower, action.toLowerCase())) {
          augmentations.push({
            conscience_value: value,
            augments: "autonomy_envelope.forbidden_actions",
            description: `${value.type} "${value.content}" reinforces forbidden action "${action}"`,
          });
        }
      }
    }

    // Check for augmentations: value reinforces escalation_triggers
    if (card.autonomy_envelope.escalation_triggers) {
      for (const trigger of card.autonomy_envelope.escalation_triggers) {
        if (containsWholeWord(contentLower, trigger.condition.toLowerCase().replace(/_/g, " ")) ||
            containsWholeWord(contentLower, trigger.condition.toLowerCase())) {
          augmentations.push({
            conscience_value: value,
            augments: "autonomy_envelope.escalation_triggers",
            description: `${value.type} "${value.content}" reinforces escalation trigger "${trigger.condition}"`,
          });
        }
      }
    }
  }

  return {
    valid: conflicts.length === 0,
    card_id: card.card_id,
    conscience_value_count: conscienceValues.length,
    conflicts,
    augmentations,
    validated_at: new Date().toISOString(),
  };
}

File: aip/packages/python/src/aip/analysis/agreement.py
"""Card-Conscience Agreement validation for the Agent Integrity Protocol.

Validates that conscience values are compatible with the Alignment Card
at initialization time. If conflicts are found, initialization MUST fail.

Port of packages/typescript/src/analysis/agreement.ts
"""

from __future__ import annotations

from datetime import datetime, timezone

from aip.schemas.agreement import (
    CardConscienceAgreement,
    CardConscienceAugmentation,
    CardConscienceConflict,
)
from aip.schemas.config import AlignmentCard
from aip.schemas.conscience import ConscienceValue


def validate_agreement(
    card: AlignmentCard,
    conscience_values: list[ConscienceValue],
) -> CardConscienceAgreement:
    """Validate that conscience values are compatible with the alignment card.

    Called at AIP initialization.

    Conflict detection:
    - A BOUNDARY value that contradicts a bounded_action (e.g., BOUNDARY says
      "never write files" but bounded_actions includes "write_files")

    Augmentation detection:
    - A BOUNDARY/FEAR value that reinforces an escalation_trigger
    - A BOUNDARY value that matches a forbidden_action
    """
    conflicts: list[CardConscienceConflict] = []
    augmentations: list[CardConscienceAugmentation] = []

    for value in conscience_values:
        # Only check BOUNDARY and FEAR types (others are not injected)
        if value.type not in ("BOUNDARY", "FEAR"):
            continue

        content_lower = value.content.lower()

        # Check for conflicts: BOUNDARY contradicts bounded_actions
        if value.type == "BOUNDARY" and card.autonomy_envelope.bounded_actions:
            for action in card.autonomy_envelope.bounded_actions:
                if (
                    action.lower().replace("_", " ") in content_lower
                    or action.lower() in content_lower
                ):
                    # Check if the BOUNDARY says "never" or "no" + the action
                    # This is a potential conflict — card allows it, conscience forbids it
                    if (
                        "never" in content_lower
                        or "no " in content_lower
                        or "don't" in content_lower
                        or "do not" in content_lower
                    ):
                        conflicts.append(
                            CardConscienceConflict(
                                conscience_value=value,
                                card_field="autonomy_envelope.bounded_actions",
                                description=(
                                    f'BOUNDARY "{value.content}" conflicts with '
                                    f'bounded action "{action}"'
                                ),
                            )
                        )

        # Check for augmentations: value reinforces forbidden_actions
        if card.autonomy_envelope.forbidden_actions:
            for action in card.autonomy_envelope.forbidden_actions:
                if (
                    action.lower().replace("_", " ") in content_lower
                    or action.lower() in content_lower
                ):
                    augmentations.append(
                        CardConscienceAugmentation(
                            conscience_value=value,
                            augments="autonomy_envelope.forbidden_actions",
                            description=(
                                f'{value.type} "{value.content}" reinforces '
                                f'forbidden action "{action}"'
                            ),
                        )
                    )

        # Check for augmentations: value reinforces escalation_triggers
        if card.autonomy_envelope.escalation_triggers:
            for trigger in card.autonomy_envelope.escalation_triggers:
                if (
                    trigger.condition.lower().replace("_", " ") in content_lower
                    or trigger.condition.lower() in content_lower
                ):
                    augmentations.append(
                        CardConscienceAugmentation(
                            conscience_value=value,
                            augments="autonomy_envelope.escalation_triggers",
                            description=(
                                f'{value.type} "{value.content}" reinforces '
                                f'escalation trigger "{trigger.condition}"'
                            ),
                        )
                    )

    return CardConscienceAgreement(
        valid=len(conflicts) == 0,
        card_id=card.card_id,
        conscience_value_count=len(conscience_values),
        conflicts=conflicts,
        augmentations=augmentations,
        validated_at=datetime.now(timezone.utc).isoformat(),
    )
